daemon.go

```diff
@@ -2,215 +2,124 @@ package main
 
 import (
 	"context"
-	"encoding/json"
+	"errors"
 	"os"
 	"os/exec"
 	"sync"
 	"time"
 
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/service/autoscaling"
-	"github.com/aws/aws-sdk-go/service/sqs"
 	log "github.com/sirupsen/logrus"
 )
 
-const (
-	heartbeatFrequency = time.Second * 10
-)
-
-type Envelope struct {
-	Type    string    `json:"Type"`
-	Subject string    `json:"Subject"`
-	Time    time.Time `json:"Time"`
-	Message string    `json:"Message"`
-}
-
-type AutoscalingMessage struct {
-	Time        time.Time `json:"Time"`
-	GroupName   string    `json:"AutoScalingGroupName"`
-	InstanceID  string    `json:"EC2InstanceId"`
-	ActionToken string    `json:"LifecycleActionToken"`
-	Transition  string    `json:"LifecycleTransition"`
-	HookName    string    `json:"LifecycleHookName"`
-}
-
+// Daemon is what orchestrates the listening and execution of the handler on a termination notice.
 type Daemon struct {
-	InstanceID  string
-	Queue       *Queue
-	AutoScaling *autoscaling.AutoScaling
-	Handler     *os.File
+	instanceID string
+	handler    Handler
+	listeners  []Listener
 }
 
-// Start the daemon.
-func (d *Daemon) Start(ctx context.Context) error {
-	if err := d.Queue.Create(); err != nil {
-		return err
+// NewDaemon creates a new Daemon.
+func NewDaemon(instanceID string, handler Handler, listeners ...Listener) *Daemon {
+	return &Daemon{
+		instanceID: instanceID,
+		handler:    handler,
+		listeners:  listeners,
 	}
+}
 
-	// ensure the queue deletion happens only once
-	var deleteOnce sync.Once
-	defer func() {
-		deleteOnce.Do(d.deleteQueue)
-	}()
-
-	if err := d.Queue.Subscribe(); err != nil {
-		return err
-	}
-	defer func() {
-		if err := d.Queue.Unsubscribe(); err != nil {
-			log.WithError(err).Error("Failed to unsubscribe from sns topic")
-		}
-	}()
+// Start the Daemon.
+func (d *Daemon) Start(ctx context.Context) error {
+	// Add a child context to stop all listeners when one has returned
+	listenerCtx, stopListening := context.WithCancel(ctx)
+	defer stopListening()
 
-	ch := make(chan *sqs.Message)
+	// Use a buffered channel to avoid deadlocking a goroutine when we stop listening
+	notices := make(chan TerminationNotice, 1)
```

Comment:
This buffered channel avoids a deadlock as long as we only have 2 listeners, but if we implemented a third listener then two of them could become deadlocked. We could use `len(d.listeners)-1` here instead?